nvse/nvse/FunctionScripts.h

```diff
@@ -130,38 +130,198 @@ class UserFunctionManager
 // allows us to call function scripts directly
 class InternalFunctionCaller : public FunctionCaller
 {
+	template <bool isSelfOwning>
+	bool PopulateArgs_Templ(ScriptEventList* eventList, FunctionInfo* info,
+		const ArrayElement_Templ<isSelfOwning>* altElemArgs);
+
 public:
 	InternalFunctionCaller(Script* script, TESObjectREFR* callingObj = NULL, TESObjectREFR* container = NULL)
 		: m_callerVersion(UserFunctionManager::kVersion), m_numArgs(0), m_script(script), m_thisObj(callingObj), m_container(container) { }
 
-	virtual ~InternalFunctionCaller() { }
-	virtual UInt8 ReadCallerVersion() {	return m_callerVersion; }
-	virtual Script * ReadScript() {	return m_script; }
-	virtual bool PopulateArgs(ScriptEventList* eventList, FunctionInfo* info);
-	virtual TESObjectREFR* ThisObj() { return m_thisObj; }
-	virtual TESObjectREFR* ContainingObj() { return m_container; }
-	
+	~InternalFunctionCaller() override = default;
+	UInt8 ReadCallerVersion() override { return m_callerVersion; }
+	Script* ReadScript() override { return m_script; }
+
+	bool PopulateArgs(ScriptEventList* eventList, FunctionInfo* info) override;
+
+	TESObjectREFR* ThisObj() override { return m_thisObj; }
+	TESObjectREFR* ContainingObj() override { return m_container; }
+
 	bool SetArgs(UInt8 numArgs, ...);
 	bool vSetArgs(UInt8 numArgs, va_list args);
 	bool SetArgsRaw(UInt8 numArgs, const void* args);
 
+	template <bool isSelfOwning>
+	bool SetArgs(UInt8 numArgs, const ArrayElement_Templ<isSelfOwning> *elemArgs);
+
 protected:
-	enum { kMaxArgs = 10 };	
 
 	UInt8			m_callerVersion;
 	UInt8			m_numArgs;
 	Script			* m_script;
-	void			* m_args[kMaxArgs];
+	void			* m_args[kMaxUdfParams];
+
+	//Points to an array of ArrElems, which retain useful type information.
+	//Used in order to remember if the nth arg was stored as double/int,
+	//since void* loses that info and forces an incorrect conversion if the UDF guessed wrong w/ having a float vs int arg.
```

Comment:
I'm referring to these lines of legacy code:
[```var->data = *((float *)&m_args[i])```
```var->data = (SInt32)(m_args[i]);```](https://github.com/xNVSE/NVSE/blob/c8eab983e4afd5a38ca5c6f48081195c41c3cdc3/nvse/nvse/FunctionScripts.cpp#L610)

If the cached args inside `InternalFunctionCaller` was always guaranteed to be double/float type in case the UDF expects a float/int script arg, there wouldn't be an issue. 

The problem is that it isn't stored as a double we can just floor; we have to guess if it's an int or a float, and if we guess wrong, the value we spit out will have an unexpected memory value due to not having the float-to-int conversion case handled (and vice-versa).

Ex: say we had stored an int. Therefore the void* for the nth arg is a pointer with the numeric value of our number.
If we then determined that the variable type for the nth arg to populate is ` Script::eVarType_Float`, then we'll be reading our stored data as:
```var->data = *((float*)&m_args[i]);```
This interprets the address of `&m_args[i]` as a `float*`, then dereferences it. But the address of our data was really `SInt32*`, and an int has a completely different memory layout then a float, so we'll populate the nth script arg with an unexpected value, instead of just promoting the int to a double.

I'm guessing that can't be easily fixed without breaking black magic used in the old `SetArgs` functions like `memcpy_s` and `va_arg(args, void *)`, so I opted to just use the `ArrayElement`s that we already have instead, since that is always known to store a double and we can convert to int if needed from there.